### ISSUE TITLE
PubSubHubbub support?

### ISSUE THREAD

**Issue by julien51** [cantino/huginn#1083]
It looks like huginn is polling feeds quite aggressively... any chance you could add support for [PubSubHubbub](http://blog.superfeedr.com/howto-pubsubhubbub/) to ease things a bit? This is a webhook based protocol which lets you be fat pinged when feeds are updated!

**Comment by cantino** [comment]
Do most RSS providers support this?

**Comment by julien51** [comment]
Well, "most" is pretty vague. Here's a list of known supporters:
* Wordpress.com
* Wordpress.org (thru plugins including default ones)
* Blogger
* Medium
* Feedburner
* FeedPress
* Superfeedr
* Etsy
* New Yorker
...

There are also countless plugins for all the popular CMSes: Drupal, ... etc

It's fairly simple to assess popularity in the RSS feeds you're following by looking at the [auto-discovery elements](http://blog.superfeedr.com/rss-autodiscovery/). Depending on the "market" for your feeds (blogs, news sites, ecommerce, ... etc) you should see an adoption rate between 10 and 50%.

I hope this helps!

**Comment by fatcatfan** [comment]
I've been investigating huggin and various things to replace Yahoo pipes. I've just been looking at what it would take to use superfeedr as an intermediate to feed huggin and came across this post. Notification pings from Superfeedr's PubSubHubbub hub can be delivered as JSON, so it seems like it could already work well with the existing huginn webhook agent. The subscribe/unsubscribe requests/responses (http://pubsubhubbub.github.io/PubSubHubbub/pubsubhubbub-core-0.4.html#rfc.section.5.3) would be the main things that need to be added and might warrant a unique agent.

**Comment by cantino** [comment]
If someone would like to contribute a PubSubHubbub agent, we'd be very open to that.

**Comment by fatcatfan** [comment]
So I'm looking at what I could do to implement this agent. Per the standard, the payload of the ping from the hub to subscriber should match the format of the subscribed content (HTML for a webpage, XML for RSS, etc.). There's no reason to replicate code for processing RSS in a potential PuSubHubbub agent, so would it make sense to add a mode to the RSS agent that can receive input from an event rather the normal mode of input from a feed URL?

**Comment by cantino** [comment]
I'd probably add it as a new Agent, since I think it'll be pretty complex and different.  But you're correct that the RSSAgent could implement [receive_web_request](https://github.com/cantino/huginn/wiki/Creating-a-new-agent#receiving-web-requests), that could be interesting too.

**Comment by fatcatfan** [comment]
I was thinking rather that a new PubSubHubbub agent would implement all the necessary logic for subscribing and receive_web_request, then emit the data it receives as an event that could be passed to RSSAgent (or other agents as appropriate to the data type). So RSSAgent would need to implement "receive".

This is similar to the request for WebsiteAgent to [extract from event payload instead of url](https://github.com/cantino/huginn/issues/1050)

**Comment by fatcatfan** [comment]
So I see there's the Commander Agent which can trigger/modify other agents. The PubSubHubbub agent could use the same methods to run an RSSAgent in response to successful call on receive_web_request. I think it could also change the target RSSAgent's schedule to "Never" since it would no longer need to run periodically. And could possibly read the feed URL from the target RSSAgent's options, rather than needing to be duplicated by the user.

**Comment by cantino** [comment]
Would the PubSubHubbubAgent be emitting full RSS feeds?  Otherwise, I don't think they make sense to send to the RSSAgent via events.

**Comment by fatcatfan** [comment]
Until I have it running I can't say with certainty what the hub sends to the webhook, but the [example shown on superfeedr's site](http://blog.superfeedr.com/howto-pubsubhubbub/) shows a "full" RSS feed XML document, it just happens to have only a single <entry> element, the new one.

**Comment by julien51** [comment]
So, the protocol says that the notification should only include:
* the new entries (there can be more than one)
* the feed-level elements (title... etc)

**Comment by fatcatfan** [comment]
I've put together a [framework for the agent](https://github.com/fatcatfan/huginn/blob/pubsubhubbub-agent/app/models/agents/pubsubhubbub_subscriber_agent.rb). It will probably be evident that I've never worked in Ruby before, and this is not working code so far as I know, just WIP as I climb this learning curve.

**Comment by fatcatfan** [comment]
"Do most RSS providers support this?" At least with superfeedr's hub, I don't think it matters. They do the polling and use some other methods to identify updates even to feeds that don't employ the PubSubHubbub methodology.

**Comment by knu** [comment]
Speaking of PubSubHubbub, a DataOutputAgent could optionally publish updates to hubs in receive().

I often create a personal feed with Huginn and subscribe to it on Feedly, but Feedly will not crawl it so often without pinging the hubs when it has only one subscriber. :disappointed:

**Comment by fatcatfan** [comment]
@knu That *would* be nice. Google runs this hub https://pubsubhubbub.appspot.com/ that could very easily be used when publishing personal feeds. Doesn't even require this Subscriber agent that I'm working on.

**Comment by fatcatfan** [comment]
@julien51 From the point of view of a subscriber, what is the difference between the open https://pubsubhubbub.superfeedr.com/ and the subscription-required https://push.superfeedr.com ? Certainly it seems the latter has a lot of extra features. Does the first one support *any* feed, or only those that actually support PuSH and publish to it?

**Comment by julien51** [comment]
@fatcatfan Well, https://pubsubhubbub.superfeedr.com/  is exactly like https://pubsubhubbub.appspot.com/ : it's a community hub which any publisher can use to publish their content. Subscribers to their content can then use this hub to get notification without requiring a Superfeedr susbcriber account.
https://push.superfeedr.com  is for superfeedr customers and works with *any feed* whether the publisher pings us or not. It has a lot of extra features (as you've noticed), but it's also a paid (with a free tier) service.

**Comment by cantino** [comment]
I believe @knu added some PubSubHubbub support to the DataOutputAgent.  I'm going to close this, but feel free to reopen to discuss further.

**Comment by sonphnt** [comment]
Hi @julien51  @cantino 

I am using https://pubsubhubbub.appspot.com/ as a hub. 
My feed has included "hub" "rel" links to this hub and located at for example http://mywerbsite.com/rss.xml

All subscribers such as Inoreader or Feedly try to add my feed URL. I dont understand how the hub can know all my subscribers to notify once there are new updates.

I tried to do a 'publish' by using this tool when I have published a news
https://pubsubhubbub.appspot.com/publish

But InoReader (setting even said 'realtime') or Feedly dont seem to get notified until i manipulate a reload by pressing refresh button.

Thanks
